core/src/main/scala/org/apache/spark/deploy/ExecutorDelegationTokenUpdater.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy
+
+import java.util.concurrent.{Executors, TimeUnit}
+import java.util.{Comparator, Arrays}
+
+import com.google.common.primitives.Longs
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{PathFilter, FileStatus, Path, FileSystem}
+import org.apache.hadoop.security.{Credentials, UserGroupInformation}
+
+import org.apache.spark.util.Utils
+import org.apache.spark.{Logging, SparkConf}
+
+private[spark] class ExecutorDelegationTokenUpdater(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration) extends Logging {
+
+  @volatile private var lastCredentialsFileSuffix = 0
+
+  private lazy val delegationTokenRenewer =
+    Executors.newSingleThreadScheduledExecutor(
+      Utils.namedThreadFactory("Delegation Token Refresh Thread"))
+
+  // On the executor, this thread wakes up and picks up new tokens from HDFS, if any.
+  private lazy val executorUpdaterRunnable =
+    new Runnable {
+      override def run(): Unit = Utils.logUncaughtExceptions(updateCredentialsIfRequired())
+    }
+
+  def updateCredentialsIfRequired(): Unit = {
+    try {
+      sparkConf.getOption("spark.yarn.credentials.file").foreach { credentialsFile =>
```

Comment:
(Actually, it's a lazy val. I hate lazy vals, so confusing.)

Anyway, I still think it would be better to have this as an `Option` on the caller site and avoid these weird "no-op by accident" semantics.